perf/evp_fetch.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/kdf.h>
+#include <openssl/core_names.h>
+#include "perflib/perflib.h"
+
+#define NUM_CALLS_PER_TEST         100000
+
+OSSL_TIME *times;
+
+int err = 0;
+
+static int threadcount;
+
+static OSSL_LIB_CTX *ctx = NULL;
+
+#define ARRAY_SIZE(a)                               \
+  ((sizeof(a) / sizeof(*(a))))
+
+typedef enum {
+    FETCH_MD = 0,
+    FETCH_CIPHER,
+    FETCH_KDF,
+    FETCH_MAC,
+    FETCH_RAND
+} fetch_type_t;
+
+struct fetch_data_entry {
+    fetch_type_t ftype;
+    const char *alg;
+    const char *propq;
+};
+
+static struct fetch_data_entry fetch_entries[] = {
+    {FETCH_MD, OSSL_DIGEST_NAME_SHA2_224, NULL},
+    {FETCH_MD, OSSL_DIGEST_NAME_SHA2_256, NULL},
+    {FETCH_MD, OSSL_DIGEST_NAME_SHA3_224, NULL},
+    {FETCH_MD, OSSL_DIGEST_NAME_SHA3_256, NULL},
+    {FETCH_CIPHER, OSSL_CIPHER_NAME_AES_128_GCM_SIV, NULL},
+    {FETCH_CIPHER, OSSL_CIPHER_NAME_AES_192_GCM_SIV, NULL},
+    {FETCH_CIPHER, OSSL_CIPHER_NAME_AES_256_GCM_SIV, NULL},
+    {FETCH_KDF, OSSL_KDF_NAME_HKDF, NULL},
+    {FETCH_KDF, OSSL_KDF_NAME_SCRYPT, NULL},
+    {FETCH_KDF, OSSL_KDF_NAME_KRB5KDF, NULL},
+    {FETCH_KDF, OSSL_KDF_NAME_KBKDF, NULL},
+    {FETCH_MAC, OSSL_MAC_NAME_BLAKE2BMAC, NULL},
+    {FETCH_MAC, OSSL_MAC_NAME_CMAC, NULL},
+    {FETCH_MAC, OSSL_MAC_NAME_GMAC, NULL},
+    {FETCH_MAC, OSSL_MAC_NAME_HMAC, NULL},
+    {FETCH_MAC, OSSL_MAC_NAME_POLY1305, NULL},
+    {FETCH_RAND, "CTR-DRBG", NULL}
+};
+
+void do_fetch(size_t num)
+{
+    OSSL_TIME start, end;
+    size_t i, j;
+
+    start = ossl_time_now();
+
+    for (i = 0; i < NUM_CALLS_PER_TEST / threadcount; i++) {
```

Comment:
The scaling here will introduce bias if `NUM_CALLS_PER_TEST` is not divisible by `threadcount`.  Specifically, the computation on line 199 doesn't account for any rounding from the integer division.

Suggest running the full `NUM_CALLS_PER_TEST` for each thread and compensating when averaging or change `(double)NUM_CALLS_PER_TEST` to `(double)((NUM_CALLS_PER_TEST / threadcount) * threadcount)` down line line 199.

Reply:
@paulidale That would make some tests that use the same construct to run for overly long time if high number of threads is used. Is this really a problem? We can just use threadcounts that are dividing 100000 without a reminder. I.e. in the original benchmark we used 1, 10, 100, 500 and 1000 as threadcount.

Reply:
Both good points.  A happy medium might be to do the division among threads as we are currently doing, but when we do the averaging, instead of dividing by NUM_CALLS_PER_TEST, instead divide by (NUM_CALLS_PER_TEST / threadcount) * threadcount).  that would adjust the final average by the actual number of iterations performed, rather than the assumed NUM_CALLS_PER_TEST.

Would that be a reasonable approach?

Reply:
Yes, that's the second option I offered and it's reasonable IMO.

Reply:
done